src/lib/utils.py

```diff
@@ -386,3 +386,33 @@ def derive_localities(localities: DataFrame, data: DataFrame) -> DataFrame:
     locs["key"] = locs["locality"]
     index_columns = ["key", "date"] if "date" in data.columns else ["key"]
     return locs.groupby(index_columns).agg(agg_func)
+
+
+def backfill_cumulative_fields(data: DataFrame, columns=[]) -> DataFrame:
+    """
+    Given a dataframe and some names of cumulative column fields,
+    backfill missing data per key.
+
+    If no columns are provided, then do all columns that begin with "total_"
+
+    Assumes both "key" and "date" are present in data.columns.
+    """
+
+    columns = columns or [col for col in data.columns if col.startswith("total_")]
+    if not columns:
+        # Early return, nothing to do here.
+        return data
+
+    groups = data.groupby(["key"])
+    new_data = data.copy()
```

Comment:
Can you avoid a full copy? This can be very memory-intensive and slow for big data sources

Reply:
I've avoided a full copy and modified the input array instead.